### PR TITLE
Chain-able includes

### PR DESCRIPTION
**Pull request by pape-io**
Minor Version

### Before
`includes` and `references` have been unchainable. Furthermore includes was implemented with `class clone` (bad practice). Not being able to chain includes and references was especially annoying when authenticating requests.

```ruby
Claims.includes(:localch_account, :entry).where(place_id: 'huU90mB_6vAfUdVz_uDoyA')
Claims.where(place_id: 'huU90mB_6vAfUdVz_uDoyA').includes(:localch_account, :entry) # raises error
```

### Now
`includes` and `references` are now fully chain-able and it's implementation is not based on suspicious class cloning any more 🎉 .

```ruby
Claims.includes(:localch_account, :entry).where(place_id: 'huU90mB_6vAfUdVz_uDoyA')
Claims.where(place_id: 'huU90mB_6vAfUdVz_uDoyA').includes(:localch_account, :entry)
authenticated(Claims)
  .where(place_id: 'huU90mB_6vAfUdVz_uDoyA')
  .includes(:localch_account, :entry)
```

### CODE REVIEW

File: lib/lhs/concerns/complex.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Complex is the main data structure for includes.
+# can be for example :place, [:place, :customer],
+# { places: [:customer] }, { places: { customer: :contract }}
+# and so on
+class LHS::Complex
```
Comment (fpellanda) [inline lib/lhs/concerns/complex.rb]: Wondering why you used a class here. This is some sort of helper module to include?

What do the methods do? I couldn't find a complex_spec.rb...

**Comment by pape-io** [comment]
@fpellanda true. Will add a test for the chain-ability. thx.

File: lib/lhs/concerns/complex.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def self.merge_datum(datum, data)
+    incompatible = datum.is_a?(Array) && data.is_a?(Hash) ||
+      datum.is_a?(Hash) && data.is_a?(Array)
+    raise "Can't merge Hash with Array" if incompatible
+    if datum.is_a?(Hash)
+      data.deep_merge! datum
+    else
+      data.push datum
```
Comment (Averethel) [inline lib/lhs/concerns/complex.rb]:
```ruby
x = [:a, :b, :c]
y = [1 ,2 ,3]

x.push(y)
=> [:a, :b, :c, [1, 2, 3]]
```

Are you sure this is intended? (opposed to `x.push(*y)`)

Anyhow it's not really obvious what's the intention here  => spec would help IMO.

File: lib/lhs/concerns/record/chainable.rb [hunk 3]
```diff
@@ -241,6 +287,18 @@ def chain_pagination
         resolve_pagination _links.select { |link| link.is_a? Pagination }
       end
 
+      def chain_includes
+        LHS::Complex.merge _links
+          .select { |link| link.is_a?(Include) && link.data.present? }
```
Comment (Averethel) [inline lib/lhs/concerns/record/chainable.rb]: `LHS::Complex.merge(_links)`

File: lib/lhs/concerns/record/chainable.rb [hunk 4]
```diff
@@ -241,6 +287,18 @@ def chain_pagination
...
+          .map { |link| link.data }
+      end
+
+      def chain_references
+        LHS::Complex.merge _links
```
Comment (Averethel) [inline lib/lhs/concerns/record/chainable.rb]: LHS::Complex.merge(_links)

File: .rubocop.yml [hunk 5]
```diff
@@ -6,6 +6,9 @@ inherit_from:
 Lint/IneffectiveAccessModifier:
   Enabled: false
 
+Lint/ShadowingOuterLocalVariable:
+  Enabled: false
```
Comment (pape-io) [inline .rubocop.yml]: For me, it just pops up in the cases where I want shadowing! it never protects me from shadowing outer local variables by accident. Anyway, usually our code is already structured in a way, with really small blocks and closures, where it's pretty unlikely to have conflicting local variables anyway.

File: lib/lhs/concerns/complex.rb (hunk 2 above)
Comment (pape-io) [inline lib/lhs/concerns/complex.rb]: Let me finish the specs. I already changed some logic here, as it was not "complete". I would be glad if you could check it again if i've finished the spec im working on right now.

File: lib/lhs/concerns/record/chainable.rb (hunk 3 above)
Comment (pape-io) [inline lib/lhs/concerns/record/chainable.rb]: ?! what @Averethel 
I want to merge the selected and map link data, not the links!

File: lib/lhs/concerns/record/chainable.rb (hunk 4 above)
Comment (pape-io) [inline lib/lhs/concerns/record/chainable.rb]: ?! what @Averethel 
I want to merge the selected and map link data, not the links!

File: lib/lhs/concerns/record/chainable.rb (hunk 3 above)
Comment (Averethel) [inline lib/lhs/concerns/record/chainable.rb]: Oh I totally misinterpreted it then. One more reason to add outer parens....

File: .rubocop.yml (hunk 5 above)
Comment (Averethel) [inline .rubocop.yml]: You're doing most development here so it's your decision. It saved my butt couple of times that's why I'm mentioning it ;)

File: lib/lhs/complex.rb [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    return nil if array.blank?
+    array
+  end
+  private_class_method :reduce!
+
+  def self.reduce_symbols!(array)
+    array.each do |element|
+      if element.is_a?(Symbol)
+        array.each do |other_element|
+          array.delete(element) if other_element.is_a?(Hash) && other_element.key?(element)
```
Comment (Averethel) [inline lib/lhs/complex.rb]: That's deleting from the array you currently iterate on (twice). I'd recommend making a copy of the array and deleting stuff from there.

File: lib/lhs/complex.rb [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    return [addition] if base.empty?
+    base.each do |element|
+      merge_multi_level!(base, element, addition)
+    end
+  end
+  private_class_method :merge_first_level!
+
+  # Merge nested complex (addition) into
+  # an hash or an array (element) of the parent
+  def self.merge_multi_level!(parent, element, addition, key = nil)
```
Comment (Averethel) [inline lib/lhs/complex.rb]: Some comment on the merging algorithm would be welcome (with examples). It's quite difficult to follow even though naming is really verbose.

File: lib/lhs/complex.rb (hunk 6 above)
Comment (pape-io) [inline lib/lhs/complex.rb]: Discussed in person

File: lib/lhs/complex.rb (hunk 7 above)
Comment (pape-io) [inline lib/lhs/complex.rb]: Discussed in person

**Pull request merged**